contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
-- [Upgrade/downgrade procedure](#upgradedowngrade-procedure)
-- [Orphaning/adoption](#orphaningadoption)
-- [Implementation plan (sketch)](#implementation-plan-sketch)
-- [Considered alternatives](#considered-alternatives)
-
-# Goal of ControllerReference
-
-Main goal of `ControllerReference` effort is to solve a problem of overlapping controllers that fight over some resources (e.g. `ReplicaSets` fighting with `ReplicationControllers` over `Pods`), which cause serious [problems](https://github.com/kubernetes/kubernetes/issues/24433) such as exploding memory of Controller Manager.
-
-We don't want to have (just) an in-memory solution, as we don’t want a Controller Manager crash to cause massive changes in object ownership in the system. I.e. we need to persist the information about "owning controller".
-
-Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
-
-# Non goals
-
-Cascading deletion is not a goal of this effort. Cascading deletion will use `ownerReferences`, which is a [separate effort](garbage-collection.md).
-
-`ControllerRef` will extend `OwnerReference` and reuse machinery written for it (GarbageCollector, adoption/orphaning logic).
-
-# API and semantic changes
-
-There will be a new API field in the `OwnerReference` in which we will store an information if given owner is a managing controller:
-
-```
-OwnerReference {
-    …
-    Controller bool
+* [Goals](#goals)
+* [Non-goals](#non-goals)
+* [API](#api)
+* [Behavior](#behavior)
+* [Upgrading](#upgrading)
+* [Implementation](#implementation)
+* [Alternatives](#alternatives)
+
+# Goals
+
+* The main goal of ControllerRef (controller reference) is to solve the problem
+  of controllers that fight over controlled objects due to overlapping selectors
+  (e.g. a ReplicaSet fighting with a ReplicationController over Pods because
+  both controllers have label selectors that match those Pods).
+  Fighting controllers can [destabilize the apiserver](https://github.com/kubernetes/kubernetes/issues/24433),
+  [thrash objects back-and-forth](https://github.com/kubernetes/kubernetes/issues/24152),
+  or [cause controller operations to hang](https://github.com/kubernetes/kubernetes/issues/8598).
+
+  We don't want to have just an in-memory solution because we don't want a
+  Controller Manager crash to cause a massive reshuffling of controlled objects.
+  Therefore, we need to persist the mapping from each object to its controller.
+
+* A secondary goal of ControllerRef is to provide back-links from a given object
+  to the controller that manages it, which can be used for:
+  * Efficient object->controller lookup, without having to list all controllers.
+  * Generic object grouping (e.g. in a UI), without having to know about all
+    third-party controller types in advance.
+  * Replacing many if not all uses of the `kubernetes.io/created-by` annotation.
```

Comment:
Perhaps a non-goal? Not sure what the created-by annotation is used for or by whom.